samples/bluetooth/observer/prj_extended.conf

```diff
@@ -0,0 +1,17 @@
+CONFIG_BT=y
+CONFIG_BT_OBSERVER=y
+
+# Enable Extended Scanning
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_EXT_SCAN_BUF_SIZE=1650
+
+# Zephyr Bluetooth LE Controller needs 16 event buffers to generate Extended
+# Advertising Report for receiving the complete 1650 bytes of data
+CONFIG_BT_BUF_EVT_RX_COUNT=16
```

Comment:
Is this really true? 

On a single core board, I would expect it to be something like

1. Controller generates event
2. Host handles event (in same thread)
3. Host releases memory for event
4. Go to 1. 

And for the dual core boards (e.g. NRF5340) the host should be able to handle the events as they come, right? 

If this is indeed the case (i.e. this value needs to be 16 to handle 1650 octets of data), then perhaps the default value should be changed? 

We have `CONFIG_BT_EXT_SCAN_BUF_SIZE` which determines the size of the events we can handle in the host. Since each report may be up to 229 octets, the number of events we need to handle should thus be `CONFIG_BT_EXT_SCAN_BUF_SIZE / 229`. 

So in case that we cannot handle each event as they come, does `CONFIG_BT_BUF_EVT_RX_COUNT` really need to be larger than 8? The default is `10` unless `CONFIG_BT_RECV_BLOCKING=y (and in which case, that should also work, right?).

Reply:
> Is this really true?

This is true. In the *best*  case,  ceil(1650 / 229) = 8 HCI reports are generated in one go by the Controller's Rx thread and while the host is recombining these 8 reports, to keep the continuity (not block the Rx thread) of receiving the reports at higher rate, 8 additional buffer for the report are allocated to realize double buffering. I.e. while 8 reports are being recombined by host, 8 new reports can be populated by the Controller's Rx Thread without deadlocking the same Rx thread that is recombining.

> And for the dual core boards (e.g. NRF5340) the host should be able to handle the events as they come, right?

Yes, for the host (application core) only sample `CONFIG_BT_BUF_EVT_RX_COUNT=1` should suffice to receive the reports from the transport, but on the Controller (network core) `CONFIG_BT_BUF_EVT_RX_COUNT=16` is required to buffer the reports until the reports are transmitted out over the transport (RPMSG/UART/SPI etc.)

> If this is indeed the case (i.e. this value needs to be 16 to handle 1650 octets of data), then perhaps the default value should be changed?

In a *worst* case a peer can send chain PDUs of 1 byte AD data, Controller does not support recombining the PDUs into report. In a typical usecase only one AUX_ADV_IND PDU is used (upto 191 byte AD data) with no chaining, under such cases the Kconfig default `CONFIG_BT_BUF_EVT_RX_COUNT=3` (under `CONFIG_BT_RECV_BLOCKING=y` is sufficient (only 1 buffer is really needed).

> So in case that we cannot handle each event as they come, does CONFIG_BT_BUF_EVT_RX_COUNT really need to be larger than 8? The default is 10 unless `CONFIG_BT_RECV_BLOCKING=y (and in which case, that should also work, right?).

As explained in first reply, in order to keep the Controller Rx Thread from blocking on allocating report buffers while the 8 reports are being recombined, 8 additional report buffers are required which correspond to the new report that is being prepared from the PDUs that have arrived on `CONFIG_BT_CTLR_RX_BUFFERS=9` (9 because Controller needs to receive the ADV_EXT_IND + AUX_ADV_IND + 7 AUX_CHAIN_IND PDUs).

Reply:
I still think I am missing some understanding here. 

In the case of `CONFIG_BT_RECV_BLOCKING=y`, why does the controller need to generate multiple events in one go? Wouldn't the controller just create a single event per advertisement it receives over air, and forward that to the host (which shouldn't require more than 1 event as it is done in the same thread). 

In the case of `CONFIG_BT_RECV_BLOCKING=n` (with the default event count of 10) the host should be able handle the events as they come, right? You even say that yourself: 

> Yes, for the host (application core) only sample CONFIG_BT_BUF_EVT_RX_COUNT=1 should suffice to receive the reports from the transport, but on the Controller (network core) CONFIG_BT_BUF_EVT_RX_COUNT=16 is required to buffer the reports until the reports are transmitted out over the transport (RPMSG/UART/SPI etc.)

Is this configuration file intended to be used on the network core too? Otherwise the increase in `CONFIG_BT_BUF_EVT_RX_COUNT` here doesn't help the network core.

Reply:
> why does the controller need to generate multiple events in one go?

The controller implementation (choice) buffers received PDU until it can decide if the last report would have a "data complete", or "data incomplete, no more to come" as data status, hence all the reports are generated in one go at the end of the reception of the chain.

Also, a single PDU that can be > 229 bytes of AD data will fragment into two advertising report, `CONFIG_BT_BUF_EVT_RX_COUNT=1` will deadlock the Rx thread.

In the case of periodic advertising reports, the implementation is different where reports are generated as the PDU arrives. This is different here as we had to interleave the IQ reports when direction finding was supported and used. Periodic Advertising report generated has in fact gotten complicated, as a additional/extra dummy report is generated with "data complete" or "data incomplete, no more to come" based on whether there will be more PDUs or the peer stops anymore PDU transmission (or fail to receive the subsequent PDUs).

Reply:
> The controller implementation (choice) buffers received PDU until it can decide if the last report would have a "data complete", or "data incomplete, no more to come" as data status, hence all the reports are generated in one go at the end of the reception of the chain.

Ah, that explains it (missed that part). 
So the controller will keep PDUs until all is received? I assume that it then aborts it if the number of PDUs (e.g. 100 PDUs of size 1 for a total of 100 octets) is larger than `CONFIG_BT_BUF_EVT_RX_COUNT`? 

Does it also handle the case where the advertising data is fragmented into e.g. 5 PDUs, P1, P2, ..., P5, and where e.g. P5 is lost, e.g.
-> P1
-> P2
-> P3
-> P4 
-/> P5 (lost)
-> P1 (should abort previously received P1 - P4 PDUs
?